util/genrest/resttools/error_response.go

```diff
@@ -62,18 +76,76 @@ func GRPCToHTTP(c codes.Code) int {
 	return httpStatus
 }
 
-// ErrorResponse is a helper that formats the given response information,
-// including the HTTP Status code, a message, and any error detail types, into
-// a googleAPIError and writes the response as JSON.
-func ErrorResponse(w http.ResponseWriter, status int, message string, details ...interface{}) {
-	googleAPIError := &googleAPIError{
-		Error: &googleapi.Error{
-			Code:    status,
+// HTTPToGRPC maps the given HTTP status code to the canonical gRPC
+// Code as defined by the internal source go/http-canonical-mapping.
+func HTTPToGRPC(httpStatus int) codes.Code {
+	var gRPCCode codes.Code
+	switch {
+	case httpStatus >= 200 && httpStatus < 300:
+		gRPCCode = codes.OK
+	case httpStatus >= 300 && httpStatus < 400:
+		gRPCCode = codes.Unknown
+	case httpStatus >= 400 && httpStatus < 500:
+		gRPCCode = codes.FailedPrecondition
+	case httpStatus >= 500 && httpStatus < 600:
+		gRPCCode = codes.Internal
+	default:
+		gRPCCode = codes.Unknown
+	}
+
+	if codeOverride, ok := httpToGRPC[httpStatus]; ok {
+		gRPCCode = codeOverride
+	}
+	return gRPCCode
+}
+
+// Constants to make it obvious when we're not supplying either a gRPC or HTTP status code to
+// ErrorResponse(). The code we're not supplying will be obtained from the one we do supply.
+const (
+	NoCodeGRPC codes.Code = 9999
+	NoCodeHTTP int        = -1
+)
+
+// ErrorResponse is a helper that formats the given response information, including the HTTP or gRPC
+// status code, a message, and any error detail types, into a RestError proto message and writes the
+// response as JSON.
+func ErrorResponse(w http.ResponseWriter, httpResponseCode int, grpcStatus codes.Code, message string, details ...interface{}) {
+	if httpResponseCode == NoCodeHTTP {
+		if grpcStatus == NoCodeGRPC {
+			WriteShowcaseRESTImplementationError(w, "neither HTTP code or gRPC status are provided for ErrorResponse. Exactly one must be provided.")
+			return
+		}
+		httpResponseCode = GRPCToHTTP(grpcStatus)
+	} else {
+		if grpcStatus != NoCodeGRPC {
+			WriteShowcaseRESTImplementationError(w, "both HTTP code and gRPC status are provided for ErrorResponse. Exactly one must be provided.")
+			return
+		}
+		grpcStatus = HTTPToGRPC(httpResponseCode)
```

Comment:
I personally prefer to avoid nesting and make conditions a bit more explicit for easier readability:
```suggestion
	if httpResponseCode == NoCodeHTTP && grpcStatus == NoCodeGRPC {
		WriteShowcaseRESTImplementationError(w, "neither HTTP code or gRPC status are provided for ErrorResponse. Exactly one must be provided.")
		return
	} else if httpResponseCode != NoCodeHTTP && grpcStatus != NoCodeGRPC {
		WriteShowcaseRESTImplementationError(w, "both HTTP code and gRPC status are provided for ErrorResponse. Exactly one must be provided.")
		return
	} else if grpcStatus != NoCodeGRPC {
		httpResponseCode = GRPCToHTTP(grpcStatus)
	} else if httpResponseCode != NoCodeHTTP {
		grpcStatus = HTTPToGRPC(httpResponseCode)
```

WDYT?

Reply:
You're right, that's more Go-idiomatic and clearer. I modified the code to a version even simpler than what you suggest.